src/sync_backend.rs

```diff
@@ -112,43 +121,45 @@ impl SyncBackend for DebugSyncBackend {
 /// data.
 pub struct RetryBackend<InnerSyncBackend> {
     inner: InnerSyncBackend,
-    delay: Duration,
-    attempts: usize,
+    min_delay: Duration,
+    max_attempts: usize,
 }
 
 impl<InnerSyncBackend> RetryBackend<InnerSyncBackend> {
     /// Creates a new backend from another SyncBackend. The max_retries parameter gives the number
     /// of times the backend will try again (so given 0, it acts just as the original SyncBackend).
     /// The delay parameter provides the amount of time to wait between each upload attempt.
-    pub fn new(inner: InnerSyncBackend, max_retries: usize, delay: Duration) -> Self {
+    pub fn new(inner: InnerSyncBackend, retries: usize, min_delay: Duration) -> Self {
         Self {
             inner,
-            delay,
-            attempts: max_retries + 1,
+            min_delay,
+            max_attempts: retries + 1,
         }
     }
 }
 
 impl<InnerSyncBackend: SyncBackend> SyncBackend for RetryBackend<InnerSyncBackend> {
     fn upload(&mut self, data: UploadInfo) -> Result<UploadResponse, Error> {
-        for index in 0..self.attempts {
-            if index != 0 {
-                log::info!(
-                    "tarmac is being rate limited, retrying upload ({}/{})",
-                    index,
-                    self.attempts - 1
-                );
-                thread::sleep(self.delay);
-            }
+        for index in 1..=self.max_attempts {
             let result = self.inner.upload(data.clone());
-
             match result {
-                Err(Error::RateLimited) => {}
+                Err(Error::RateLimited(retry_after)) => {
+                    let time = max(self.min_delay, Duration::new(retry_after, 0));
+                    log::info!(
+                        "tarmac is being rate limited, retrying upload after {:?} ({} of {} retries)",
+                        time,
+                        index,
+                        self.max_attempts - 1
+                    );
+                    thread::sleep(time);
+                    if index == self.max_attempts {
+                        return Err(Error::RateLimited(retry_after));
+                    }
+                }
                 _ => return result,
             }
         }
-
-        Err(Error::RateLimited)
+        unreachable!()
```

Comment:
I think it's preferable to avoid the `unreachable!()` call here just to rely on rust checks

Reply:
We can't return Error::RateLimited because it has a required property now.

Reply:
However, this codepoint **is** really unreachable. There's two paths the for loop can take:
match result:
- `_ => return result` means the for loop does not complete.
- `Err(Error::RateLimited` etc includes `if index == self.max_attempts { return` etc, meaning that the for loop will never complete its last iteration.
In both cases, the for loop exits before completing, so the code after that never runs.